theano/gof/opt.py

```diff
@@ -530,10 +528,10 @@ def process_node(self, fgraph, node):
             if len(node.inputs) != len(candidate.inputs):
                 continue
 
-            inputs_match = all(node_in is cand_in
-                               for node_in, cand_in in zip(node.inputs,
-                                                           candidate.inputs))
-            if inputs_match and node.op == candidate.op:
+            if node.op != candidate.op:
+                continue
```

Comment:
Indeed, in that particular case, `node.op == candidate.op` _and_ `node.op != candidate.op`.
The reason is that `__eq__` is defined, but `__ne__` is never defined to return `not __eq__`, for any Op.
This was discovered because of the test of scan, but I'm suspecting that all Ops with attributes that should not be used for comparison could be affected.